components/users/user.js

```diff
@@ -1,32 +1,73 @@
 import mongoose from 'mongoose'
 import bcryptjs from 'bcryptjs'
+import jwt from 'jsonwebtoken'
 
 const userSchema = new mongoose.Schema({
   name: {
     type: String,
-    required: true
+    required: true,
   },
   email: {
     type: String,
     required: true,
-    unique: true
+    unique: true,
   },
   password: {
     type: String,
     required: true,
-    minlength: 8
-  }
+    minlength: 8,
+  },
+  tokens: [
+    {
+      token: {
+        type: String,
+        required: true,
+      },
+    },
+  ],
 })
 
-userSchema.pre('save', async function(next) {
+userSchema.pre('save', async function (next) {
   const user = this
 
-  if(!user.isModified('password')) return
+  if (!user.isModified('password')) return
 
   user.password = await bcryptjs.hash(user.password, 8)
   next()
 })
 
+userSchema.methods.generateToken = async function () {
+  const user = this
+  const secretKey = process.env.SECRET_KEY || 'secret'
+  const token = jwt.sign({ _id: user._id.toString() }, secretKey)
+  user.tokens = user.tokens.concat({ token })
+  await user.save()
+  return token
+}
+
+userSchema.statics.findUser = async (email, password) => {
+
+  const user = await User.findOne({ email });
+
+  const errorMessage = 'Unable to login';
+
+  if(!user) throw new Error(errorMessage)
+
+  const passwordMatch = await bcryptjs.compare(password, user.password);
+  if(!passwordMatch) throw new Error(errorMessage)
+
+  return user
+}
+
+userSchema.methods.toJSON = function() {
+  const user = this;
+  const userObject = user.toObject()
+  delete userObject.password
+  delete userObject.tokens
+  delete userObject._id
+  return userObject
+}
```

Comment:
Can you explain the use of this method? we are just deleting most of the properties

Reply:
I simplified this method, the idea is to select the user properties that we want to send, which is like a mapper.